Optimize TSM compaction

jwilder opened the pull request:
This PR optimizes the TSM compactions so that they can keep up with the cache compactions.  It does this re-working the `KeyIterator` interface to return encoded blocks instead of `[]Value` which allows the TSM compaction to skip decoding for consecutive blocks that don't overlap in time.  Another change is that the compaction process now just iterates over each block in order instead of searching the index and reading all the values for a key.  This also reduces the memory footprint for larger compactions.

Some other changes that are part of the PR are:

* Optimized TSM file loading to not allocate as much when initializing the `indirectIndex`
* Load TSM files concurrently when the file store is opened
* Fixed a bug where TSM files would be out of order in the file store causing incorrect query results
* Fixed a decoding bug in `influx_inspect` that would report string blocks as corrupt
* Added max value block size chunking back (was removed when `MergeIterator` was deleted)

There are more optimizations that could be made to avoid decoding blocks in some cases, but have left those for another time since this PR seems to resolve the TSM compaction slowness as is.

@otoolep @pauldix

otoolep commented on tsdb/engine/tsm1/file_store.go:
```diff
@@ -230,15 +239,28 @@ func (f *FileStore) Open() error {
-		if err != nil {
-			return fmt.Errorf("error opening memory map for file %s: %v", fn, err)
-		}
+		go func(file *os.File) {
```
How big a difference did this make for you?

otoolep commented on tsdb/engine/tsm1/compact.go:
```diff
@@ -301,7 +303,7 @@ func (c *Compactor) Compact(tsmFiles []string) ([]string, error) {
 	}
 
-	tsm, err := NewTSMKeyIterator(trs...)
+	tsm, err := NewTSMKeyIterator(tsdb.DefaultMaxPointsPerBlock, trs...)
```
I will continue to review this code, but I want to check that this doesn't mean that this iterator will have any problems working with files that were created with a different value for `DefaultMaxPointsPerBlock`.

jwilder replied:
Yes.  That is the size of blocks that it will emit.  It doesn't matter what the block sizes are within the files it that it reads.

otoolep commented on tsdb/engine/tsm1/file_store.go:
```diff
@@ -237,18 +244,34 @@ func (f *FileStore) Open() error {
+	for range files {
+		res := <-readerC
+		if res.err != nil {
+
```
Nit: blank line looks superfluous.

otoolep commented on tsdb/engine/tsm1/writer.go:
```diff
@@ -0,0 +1,632 @@
+	return nil
+}
+
+func (t *tsmWriter) Write(key string, values Values) error {
```
Does it matter if `len(values)` is nil?

jwilder replied:
I'll add a test to handle this case.

otoolep commented on tsdb/engine/tsm1/reader.go:
```diff
@@ -1323,6 +872,17 @@ func (m *mmapAccessor) readBlock(entry *IndexEntry, values []Value) ([]Value, er
+		return nil, ErrTSMClosed
+	}
+
+	return m.b[entry.Offset+4 : entry.Offset+int64(entry.Size)], nil
```
I'd like to suggest a comment that this `+4` is about the CRC, just so it's clear.

jwilder replied:
Done.

pauldix commented:
Lot here, but looks good. I assume the answer is yes, so please excuse the dumb question, but is there a test that combines a few TSM files where one has a block that can just get written but then some other ones have blocks that have to be decoded and sorted?

Do you have a benchmark data on some of this is it just through stress running locally on your laptop?

I'm curious what the speedup is in terms of reading the indexes across many files with many series and what the speedup is for compaction.

otoolep commented on cmd/influx_inspect/tsm.go:
```diff
@@ -550,14 +550,13 @@ func cmdDumpTsm1dev(opts *tsdmDumpOpts) {
 			f.Read(b[:4])
 
-			chksum := btou32(b)
+			chksum := btou32(b[:4])
```
Can we put a comment here?

otoolep commented:
Lots here is right, I took a first pass. I'll take a second, but I'm OK with this going in.

otoolep commented on tsdb/engine/tsm1/compact.go:
```diff
@@ -443,110 +445,188 @@ type tsmKeyIterator struct {
-		// Grab the key for this reader
-		key, entries := r.Key(k.pos[i])
-		k.keys[i] = key
+	// Now we need to find all blocks that match the min key so we can combine and dedup
```
dedup -> dedupe

otoolep commented on tsdb/engine/tsm1/writer.go:
```diff
@@ -0,0 +1,632 @@
+	return fmt.Sprintf("min=%s max=%s ofs=%d siz=%d", e.MinTime.UTC(), e.MaxTime.UTC(), e.Offset, e.Size)
+}
+
+func NewDirectIndex() TSMIndex {
```
What's your plan for removing direct-indexing code?

jwilder replied:
It can't be removed.  It's needed by the TSMWriter.  The readers only use indirect indexing now though.

otoolep replied:
OK, for some reason I thought indirect would be an entire replacement for direct, but you're saying only for the read case.

jwilder commented:
@pauldix There was no test for that case because I had not implemented a fast block count that avoids decoding.  Was planning to do that in a follow up PR. I initially tried just decoding and counting, but that causes every block to be decoded too frequently which causes things to bog down again.  I updated this PR with the fast block count and added a test for it so now it will not decode blocks that are full and don't need to be de-duped.

I don't have a benchmark for reading many files right now.  I've been running various flavors of `influx_stress` tests locally and monitoring the TSM file counts.  

For startup slowness, I mainly focused on tuning the `indirectIndex.UnmarshalBinary` function because that was grossly inefficient with allocations according to the profile @otoolep captured.  I added a benchmark for that code now.  Here's the before (master) and after (this PR):

Original:
```
BenchmarkIndirectIndex_UnmarshalBinary-8	      50	  27478259 ns/op	14214973 B/op	  400001 allocs/op
```

Optimized:
```
BenchmarkIndirectIndex_UnmarshalBinary-8	    1000	   2207007 ns/op	 2021121 B/op	       2 allocs/op
```

The issue was the `400001` allocations which has been reduced to `2`.  This is with 100,000 series.  With 1M series, the old allocations increases to `4,000,014` and now they stay constant at `2` regardless of number of series.

```
name                             old time/op    new time/op    delta
IndirectIndex_UnmarshalBinary-8    27.6ms ± 1%     2.2ms ± 2%   -92.00%  (p=0.008 n=5+5)

name                             old alloc/op   new alloc/op   delta
IndirectIndex_UnmarshalBinary-8    14.2MB ± 0%     2.0MB ± 0%   -85.78%  (p=0.008 n=5+5)

name                             old allocs/op  new allocs/op  delta
IndirectIndex_UnmarshalBinary-8      400k ± 0%        0k ± 0%  -100.00%  (p=0.008 n=5+5)
```

pauldix commented:
@jwilder sounds good, anything left to get this mergeable?

jwilder commented:
Just need to fixup a few comments for @otoolep comments and then can merge.

jwilder commented:
All comments addressed.

otoolep commented:
Great stuff @jwilder

pauldix commented:
BOOYA!

side note: is there some booya emoji that we can use?
